### ISSUE TITLE
Threadsafe angle parsing

### ISSUE THREAD

**Issue by atcnofs** [astropy/astropy#7168]
The current angle parsing code is not thread safe with string inputs. This can be reproduced with:
```python
import astropy.coordinates, threading
angles = ['00:00:00']*10000
def parse_test(i=0):
   try:
      print(i, len(astropy.coordinates.Angle(angles, unit='hour')))
   except:
      print('{0} failed.'.format(i))
for i in range(10):
    threading.Thread(target=parse_test, args=(i,)).start()
```
Many of the threads should fail due to parsing errors.

The issue appears to be due to the reliance on the external ply.yacc for parsing. From the yacc header:
```python
# The current implementation is only somewhat object-oriented. The
# LR parser itself is defined in terms of an object (which allows multiple
# parsers to co-exist).  However, most of the variables used during table
# construction are defined in terms of global variables.  Users shouldn't
# notice unless they are trying to define multiple parsers at the same
# time using threads (in which case they should have their head examined).
```
yacc is also used in several places in astropy.units but I have not been able to find a reproducible failure case--I have encountered a rare error with Time parsing where a bad initialization appears to cause a failure in all time parses until Python is restarted.

Changing `_AngleParser.__init__` and `_AngleParser.parse` to use an object-level `threading.RLock` mitigates the failure:
```python
def __init__(self):
    import threading
    self.__lock = threading.RLock()
    if '_parser' not in _AngleParser.__dict__:
        _AngleParser._parser, _AngleParser._lexer = self._make_parser()

def parse(self, angle, unit, debug=False):
    try:
        with self.__lock:
            found_angle, found_unit = self._parser.parse(
                   angle, lexer=self._lexer, debug=debug)
    except ValueError as e:
        if str(e):
            raise ValueError("{0} in angle {1!r}".format(
                str(e), angle))
        else:
            raise ValueError(
                "Syntax error parsing angle {0!r}".format(angle))

    if unit is None and found_unit is None:
        raise u.UnitsError("No unit specified")

    return found_angle, found_unit
```
The description in the yacc header suggests this is not a complete solution.

**Comment by eteq** [comment]
Fascinating - thanks for the investigation @atcnofs.  Thread-safety has never been *explicitly* a goal, but it's definitely a nice-to-have.  And it really *shouldn't* be a problem here given how simple in concept the `Angle` object is.  So I think we should aim to resolve this in 3.1, but only backport it to 3.0 *if* the fix is straightforward (as opposed to a major re-work of the PLY usage or something).

**Comment by eteq** [comment]
(I am calling it a feature request and bug simultaneously based on the above, and some conversation in the Slack #coordinates channel about this)

**Comment by giacomov** [comment]
I just encountered the same problem with Units, i.e., the string parsers for the Unit class are not thread safe and I was getting weird errors. I was very puzzled at the beginning because repeating the code multiple times was giving different errors, but always related to partially or wrongly interpreted units.

While we wait for this to be fixed, I have a workaround for some applications like mine, maybe somebody could find it useful. 

This assumes that you can decide the format of your input string, i.e., most probably you are not reading it from a file. Then you can define your own format and make it thread safe. For example, this defines the "threadsafe" format (tested in python 2.7):

```python
import re
from astropy.units.format.base import Base
import astropy.units as u


# NOTE: the metaclass in Base will take care of registering
# this format, which will be available in the u.Unit
# constructor and in the .to_string method of the Unit and
# Quantity classes as format='threadsafe' (case insensitive)

def _format_one((base, power)):
    # This is for example 'cm-2' if base=cm and power=-2,
    # but only 'cm' if base=cm and power=1

    return "%s%s" % (base.to_string(), power if power != 1 else '')


class ThreadSafe(Base):

    @classmethod
    def parse(cls, s):

        # Since often we get these strings from a YAML file,
        # we want to make sure there is no new line within
        # the string.
        assert "\n" not in s

        # We assume that the string is something like
        # [unit1]{power} [unit2]{power} ..., like for example:
        # "m s-1" (a speed) or "keV-1 cm-2 s-1"
        # (a differential flux in 1 / (keV cm**2 s))
        # This is of course the format that is the output of
        # our to_string method(). See there for details

        tokens = re.findall('([a-zA-z]+)(-?\+?[0-9]+)?', s)

        # tokens is a list of tuples of the type [(unit name, power), ...]
        # Here we build a list like [u.m, u.s**(-1), ...]
        r = []

        for (unit, power) in tokens:

            # Get the primitive unit from the units module
            thisr = getattr(u, unit)

            # If needed, raise it to the power
            if power != '':

                thisr = thisr**power

            r.append(thisr)

        # Now we multiply the units
        # so that we get something like u.m * u.s**(-1)

        return reduce(lambda x, y: x*y, r)

    @classmethod
    def to_string(cls, unit):

        # The Unit class has two lists: bases (which are primitive units)
        # and powers (the correspondent powers). Thus we just need to use
        # those and encode them in a string like [unit1]{power} [unit2]{power} ...

        tokens = map(_format_one, zip(unit.bases, unit.powers))

        return str(" ".join(tokens))
```

**Comment by eteq** [comment]
I'm delaying this to 3.2 since no progress has been made.  If it's not clear: anyone who knows `yacc` that might have an alternative way to approach this please consider taking this on!

(And note that despite the yacc comment noted in  https://github.com/astropy/astropy/issues/7168#issue-295255866, if you do so we will not thing your head needs examining... rather a hand shake/hug!)

**Comment by MSeifert04** [comment]
I would rather say we should mention it in the documentation that parsing is not thread safe. The problem with locks is that they can be really tricky to get right. Race-conditions (like the one that happened here) are ugly but deadlocks (not sure if possible here - just as an example) are really annoying and locks generally introduce a lot of complexity.

The other option (besides locks) would be to use a thread-safe parser library or roll our own.

**Comment by bsipocz** [comment]
Actually I remove the milestone. If we don't have a clear path ahead it's better not just roll issues from milestone to milestone, and if actually someone comes along at any point the fix for this should end up in a bugfix release.

Also 👍 for @MSeifert04's idea of mentioning this in the docs for now.

**Comment by mhvk** [comment]
Just saw this as it drifted up the "most recently changed" list, and it left me somewhat confused: the comment in the `yacc` header states that one should not notice unless one is "trying to define multiple parsers at the same time using threads", yet the actual errors seem to occur because we are *using* the same parser in multiple threads.

Now if it were really a problem that we're actually defining a parser time and again, we should fix that.  To the extent that I understand what we do, however, it does seem like we're just re-using the parser in multiple threads (which does not seem to be something for which we should have our heads examined...). So, the problem is, I think, that the parser changes it own internal state while parsing (from a quick look, e.g., via `statestack` and `symstack`). Is this analysis correct? If so, that might also mean upstream could be willing to fix it. Or perhaps we can suggest a fix? Surely better for the larger ecosystem than "roll our own".

**Comment by MSeifert04** [comment]
Good question. I guess the first step would be to locate the race-condition.

Just a side note though: We're using 

```python
class _AngleParser:
    def __init__(self):
        if '_parser' not in _AngleParser.__dict__:
            _AngleParser._parser, _AngleParser._lexer = self._make_parser()
```

Which isn't thread-safe because in between the `if` and the assignment could run any kind of code in another thread (including another `_AngleParser.__init__`. I haven't checked if that's the problem, but it definitely would fit the comment.

**Comment by mhvk** [comment]
@MSeifert04 - I did a quick test where I ensured an angle was parsed before any threads were started - just run `parse_test()` before starting the loop with the threads, and that does not seem to be the problem: the locking seems required where @atcnofs takes the lock in his sample code, which is before doing actual parsing. I fear it is the parser itself which is keeping state while parsing, when logically that state would need to be kept in a thread-specific object. (It seems an interesting puzzle, but not sure I have the time to really look into it...)

**Comment by MSeifert04** [comment]
@mhvk You're right. It seems weird that a parser keeps state (at least global state) though. Maybe this is more of an upstream-issue?

**Comment by bnikolic** [comment]
We came across this issue. 
The root problem seems that because parser and lexer are class members (i.e., not instance members), a single parser instance is being shared between threads. So it is not a global state but rather internal state that is being corrupted. As far as I can see any solution which does not create a new parser instance for each angle object needs to take into account threading.  Given this, most effective is to use threading.local() storage for the parsers so that each thread has a separate copy but otherwise they can continue to process (pseudo) concurrently.

See MR #10556
